File: mobile/ios/OpenU60/Features/RouterSettings/SignalDetect/SignalDetectViewModel.swift
import SwiftUI

@Observable
@MainActor
final class SignalDetectViewModel {
    var status: SignalDetectStatus = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    private let client: AgentClient
    private let authManager: AuthManager
    private var pollTask: Task<Void, Never>?

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func startDetection() async {
        isLoading = true
        message = nil
        status.results = []

        do {
            let _ = try await client.postJSON("/api/cell/signal-detect/start")
            status.running = true
            showMessage("Detection started", isError: false)
            startPolling()
        } catch {
            showMessage("Failed to start: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func stopDetection() async {
        stopPolling()

        do {
            let _ = try await client.postJSON("/api/cell/signal-detect/stop")
            status.running = false
            showMessage("Detection stopped", isError: false)
            await fetchResults()
        } catch {
            showMessage("Failed to stop: \(error.localizedDescription)", isError: true)
        }
    }

    func fetchResults() async {
        do {
            let data = try await client.getJSON("/api/cell/signal-detect/results")
            status.results = SignalDetectParser.parseResults(data)
        } catch {
            // Results may not be available yet
        }
    }

    private func startPolling() {
        stopPolling()
        pollTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(2))
                await pollProgress()
                if !status.running { break }
            }
        }
    }

    private func stopPolling() {
        pollTask?.cancel()
        pollTask = nil
    }

    private func pollProgress() async {
        do {
            let data = try await client.getJSON("/api/cell/signal-detect/progress")
            let progressStatus = SignalDetectParser.parseProgress(data)
            status.progress = progressStatus.progress
            if progressStatus.progress >= 100 {
                status.running = false
                stopPolling()
                await fetchResults()
                showMessage("Detection complete", isError: false)
            }
        } catch {
            // Continue polling
        }
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/signaldetect/SignalDetectViewModel.kt
package com.openu60.feature.router.signaldetect

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.SignalDetectParser
import com.openu60.core.model.SignalDetectStatus
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SignalDetectState(
    val status: SignalDetectStatus = SignalDetectStatus.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class SignalDetectViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(SignalDetectState())
    val state: StateFlow<SignalDetectState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/modem/signal-detect/status")
                val status = SignalDetectParser.parseProgress(data)
                _state.value = _state.value.copy(status = status, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun startDetect() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/modem/signal-detect")
                // Poll until done
                while (true) {
                    delay(2000)
                    val data = agentClient.getJSON("/api/modem/signal-detect/status")
                    val progress = SignalDetectParser.parseProgress(data)
                    _state.value = _state.value.copy(status = progress)
                    if (!progress.running) break
                }
                // Fetch results
                val resultData = agentClient.getJSON("/api/modem/signal-detect/status")
                val results = SignalDetectParser.parseResults(resultData)
                _state.value = _state.value.copy(
                    status = _state.value.status.copy(results = results),
                    isLoading = false,
                    message = "Scan complete: ${results.size} results",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) startDetect() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}
